src/ring_theory/graded_algebra/radical.lean

```diff
@@ -0,0 +1,190 @@
+  -- `  proj (max₁ + max₂) (x * y)`
+  -- `= ∑ {(i, j) ∈ supports | i + j = max₁ + max₂}, xᵢ * yⱼ`
+  -- `= proj max₁ x * proj max₂ y`
+  -- `  + ∑ {(i, j) ∈ supports \ {(max₁, max₂)} | i + j = max₁ + max₂}, xᵢ * yⱼ`.
+  -- This is a contradiction, because both `proj (max₁ + max₂) (x * y) ∈ I` and the sum on the
+  -- right hand side is in `I` however `proj max₁ x * proj max₂ y` is not in `I`.
+  letI : Π (x : A),
+    decidable_pred (λ (i : ι), proj 𝒜 i x ∉ I) := λ x, classical.dec_pred _,
+  letI : Π i (x : 𝒜 i), decidable (x ≠ 0) := λ i x, classical.dec _,
+  set set₁ := (support 𝒜 x).filter (λ i, proj 𝒜 i x ∉ I) with set₁_eq,
+  set set₂ := (support 𝒜 y).filter (λ i, proj 𝒜 i y ∉ I) with set₂_eq,
+  have set₁_nonempty : set₁.nonempty,
+  { rw filter_nonempty_iff,
+    contrapose! rid₁,
+    rw ← sum_support_decompose 𝒜 x,
+    apply ideal.sum_mem _ rid₁,},
+  have set₂_nonempty : set₂.nonempty,
+  { rw filter_nonempty_iff,
+    contrapose! rid₂,
+    rw ← sum_support_decompose 𝒜 y,
+    apply ideal.sum_mem _ rid₂, },
+  set max₁ := set₁.max' set₁_nonempty with max₁_eq,
+  set max₂ := set₂.max' set₂_nonempty with max₂_eq,
+  have mem_max₁ := max'_mem set₁ set₁_nonempty,
+  have mem_max₂ := max'_mem set₂ set₂_nonempty,
+  replace hxy : (decompose 𝒜 (x * y) (max₁ + max₂) : A) ∈ I := hI _ hxy,
+  have eq :=
+    calc  proj 𝒜 (max₁ + max₂) (x * y)
+        = ∑ ij in ((support 𝒜 x).product (support 𝒜 y)).filter (λ z, z.1 + z.2 = max₁ + max₂),
+            (proj 𝒜 ij.1 x) * (proj 𝒜 ij.2 y)
+        : begin
+          rw [proj_apply, alg_equiv.map_mul, support,support, direct_sum.coe_mul_apply_submodule],
+          refl,
+        end
+    ... = ∑ ij in (((support 𝒜 x).product (support 𝒜 y)).filter
+            (λ (z : ι × ι), z.1 + z.2 = max₁ + max₂)).erase (max₁, max₂),
+            (proj 𝒜 ij.1 x) * (proj 𝒜 ij.2 y) +
+          (proj 𝒜 max₁ x) * (proj 𝒜 max₂ y)
+        : begin
+          rw sum_erase_add,
+          simp only [mem_filter, mem_product, eq_self_iff_true, and_true],
+          exact ⟨(filter_subset _ _) mem_max₁, (filter_subset _ _) mem_max₂⟩,
+        end,
+
+  have eq₂ : (proj 𝒜 max₁) x * (proj 𝒜 max₂) y
+          = proj 𝒜 (max₁ + max₂) (x * y)
+          - ∑ (ij : ι × ι) in (((support 𝒜 x).product (support 𝒜 y)).filter
+              (λ (z : ι × ι), z.fst + z.snd = max₁ + max₂)).erase (max₁, max₂),
+              (proj 𝒜 ij.fst) x * (proj 𝒜 ij.snd) y,
+  { rw [eq, eq_sub_iff_add_eq, add_comm], },
+
+  have mem_I : (proj 𝒜 max₁) x * (proj 𝒜 max₂) y ∈ I,
+  { rw eq₂,
+    refine ideal.sub_mem _ hxy (ideal.sum_mem _ (λ z H, _)),
+    rcases z with ⟨i, j⟩,
+    simp only [mem_erase, prod.mk.inj_iff, ne.def, mem_filter, mem_product] at H,
+    rcases H with ⟨H₁, ⟨H₂, H₃⟩, H₄⟩,
+    have max_lt : max₁ < i ∨ max₂ < j,
+    { rcases lt_trichotomy max₁ i with h | rfl | h,
+      { exact or.inl h },
+      { refine false.elim (H₁ ⟨rfl, add_left_cancel H₄⟩), },
+      { apply or.inr,
+        have := add_lt_add_right h j,
```

Comment:
I tried to replace `linear_ordered_cancel_add_comm_monoid` to `linear_order` and `add_left_cancel_semi_group`, but this line requires
`[_inst_3 : covariant_class α α (function.swap has_add.add) has_lt.lt]` which `linear_order` + `add_left_cancel_semi_group` cannot synthesize together

Reply:
This is probably because we still need to say `commutative \iota` somewhere, so I tried to add `linear_order` `add_left_cancel_semigroup` and `add_comm_monoid`, then some proof breaks, I think it is because lean gets confused  about `+` in `add_left_cancel_semigroup` and `+` in `add_comm_monoid`